### PR TITLE
SQL: integer parameter validation in string functions

### LINKED ISSUE #58923: SQL: integer parameter validation in string functions
The following query uses `Integer.MAX_VALUE` as parameter input: `SELECT SUBSTRING('a', 2147483647, 1), INSERT('ab', 2147483647, 1, 'x')`, returning the correct result:
```
SUBSTRING('a', 2147483647, 1)|INSERT('ab', 2147483647, 1, 'x')
-----------------------------+--------------------------------
                             |ab                              
```

Going over the Integer limit by 1:
* `SELECT SUBSTRING('a', 2147483648, 1)` returns an expected error: 
  ```
        "type" : "string_index_out_of_bounds_exception",
        "reason" : "begin 2147483647, end -2147483648, length 1"
  ```
* `SELECT INSERT('ab', 2147483648, 1, 'x')` is surprisingly accepted and returns a correct result:
  ```
  INSERT('ab', 2147483648, 1, 'x')
  --------------------------------
  ab                              
  ```
. But going over the limit with more than 1: `SELECT SUBSTRING('a', 2147483649, 1), INSERT('ab', 2147483649, 1, 'x')` behaves as an integer overflow:
```
SUBSTRING('a', 2147483649, 1)|INSERT('ab', 2147483649, 1, 'x')
-----------------------------+--------------------------------
a                            |xb                              
```

The expectation would be either for the functions to accept a long parameter _or_ throw an arithmetic overflow error.

### PR DESCRIPTION
**Pull request by bpintea**
This PR adds integer parameter validation for the string functions taking one: LOCATE, LEFT, INSERT, REPEAT, RIGHT, SUBSTRING.

Closes #58923.
Supersedes #60281.

Co-authored-by:  yinanwu <yinanwu@tencent.com>

### CODE REVIEW

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertProcessorTests.java [hunk 1]
```diff
@@ -52,33 +52,53 @@ public void testInsertWithEdgeCases() {
...
                 () -> new Insert(EMPTY, l("foobar"), l("c"), l(3), l("baz")).makePipe().asProcessor().process(null));
-        assertEquals("A number is required; received [c]", siae.getMessage());
+        assertEquals("A fixed point number is required for [start]; received [java.lang.String]", siae.getMessage());
+
         siae = expectThrows(SqlIllegalArgumentException.class,
                 () -> new Insert(EMPTY, l("foobar"), l(1), l('z'), l("baz")).makePipe().asProcessor().process(null));
-        assertEquals("A number is required; received [z]", siae.getMessage());
+        assertEquals("A fixed point number is required for [length]; received [java.lang.Character]", siae.getMessage());
+
+        assertEquals("baroobar", new Insert(EMPTY, l("foobar"), l(-2147483647), l(1), l("bar")).makePipe().asProcessor().process(null));
```
Comment (matriv) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertProcessorTests.java:84]: minor: Could you use Integer.MIN/MAX_VALUE to make it more clear?

**Review by matriv — APPROVED** [review]
LGTM. Nice work with the tests! (left minor comment)

PS: It would be great if in the future we could test all those invalid args at one place, maybe by iterating over the subclasses of the superclass of those processors, and they had as package private their accepted limits so they could be used in that test.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertProcessorTests.java (hunk 1 above)
Comment (wuyinan0126) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertProcessorTests.java:84]: @matriv OK, I will fix it

**Comment by bpintea** [comment]
@elasticmachine update branch

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertProcessorTests.java (hunk 1 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertProcessorTests.java]: [Fix](https://github.com/elastic/elasticsearch/pull/60281/commits/7d8ecfc8aece028a965c1fad017c7e137696989c) ported.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/Check.java [hunk 2]
```diff
@@ -36,4 +36,15 @@ public static void notNull(Object object, String message, Object... values) {
...
     }
+
+    public static void isFixedNumberAndInRange(Object object, String objectName, Long from, Long to) {
+        if ((object instanceof Number) == false || object instanceof Float || object instanceof Double) {
+            throw new SqlIllegalArgumentException("A fixed point number is required for [{}]; received [{}]", objectName,
+                object.getClass().getTypeName());
+        }
+        Long longValue = ((Number) object).longValue();
+        if (longValue < from || longValue > to) {
+            throw new SqlIllegalArgumentException("[{}] equals [{}], out of the allowed range [{}..{}]", objectName, longValue, from, to);
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/Check.java:47]: Unless we use this format elsewhere [{} .. {}] (with two dots), I would use the comma format [{}, {}].

Comment (astefan) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/Check.java:47]: Also, I think the style of the error message we use is first state the problem, then the actual value. Something like "[start] out of the allowed range [ ], received [{}]". Found one example (maybe there are more with out of range values) in [the Analyzer](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java#L529).

**Review by astefan — APPROVED** [review]
LGTM in general, left couple of comments regarding the wording of the error message.

**Comment by bpintea** [comment]
@elasticmachine update branch

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/Check.java (hunk 2 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/Check.java]: Good point, updated it.

**Pull request merged**